src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,204 @@
+            return TryAcquireLock(name, maxValidFor, timeout, GetMachineName(), GetThreadId(), out @lock);
+        }
+
+        public DistributedLock AcquireLockForThread(string name, TimeSpan maxValidFor, TimeSpan? timeout) {
+            return AcquireLock(name, maxValidFor, timeout, GetMachineName(), GetThreadId());
+        }
+
+        public void ReleaseLock(DistributedLock @lock) {
+            lock (_transactionManagerLock) {
+                var childLifetimeScope = CreateChildLifetimeScope(@lock.Name);
+
+                try {
+                    var repository = childLifetimeScope.Resolve<IRepository<DistributedLockRecord>>();
+                    var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                    transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+                    var lockId = Int32.Parse(@lock.Id);
+                    var record = repository.Get(lockId);
+
+                    if (record == null)
+                        throw new ObjectDisposedException("@lock", "No lock record could be found for the specified lock to be released.");
+
+                    if (record.Count <= 0)
+                        throw new ObjectDisposedException("@lock", "The specified lock has already been released.");
+
+                    record.Count--;
+                }
+                catch (Exception ex) {
+                    if (ex.IsFatal()) throw;
+                    Logger.Error(ex, "An non-fatal error occurred while trying to dispose a distributed lock with name '{0}' and ID {1}.", @lock.Name, @lock.Id);
+                }
+                finally {
+                    childLifetimeScope.Dispose();
+                }
+            }
+        }
+
+        private bool TryAcquireLock(string name, TimeSpan maxValidFor, TimeSpan? timeout, string machineName, int? threadId, out DistributedLock @lock) {
+            @lock = AcquireLockInternal(name, maxValidFor, machineName, threadId, timeout.GetValueOrDefault());
+            if (@lock != null)
+                return true;
+
+            Logger.Debug("Could not acquire a lock named '{0}'.", name);
+            return false;
+        }
+
+        private DistributedLock AcquireLock(string name, TimeSpan maxValidFor, TimeSpan? timeout, string machineName, int? threadId) {
+            var @lock = AcquireLockInternal(name, maxValidFor, machineName, threadId, timeout);
+            if (@lock != null)
+                return @lock;
+
+            throw new TimeoutException("Could not acquire a lock within the specified amount of time.");
+        }
+
+        private DistributedLock AcquireLockInternal(string name, TimeSpan maxValidFor, string machineName, int? threadId, TimeSpan? timeout = null) {
+            try {
+                DistributedLockRecord record = null;
+                var acquired = Poll(() => (record = AcquireLockRecord(name, maxValidFor, machineName, threadId)) != null, timeout);
+
+                if (acquired) {
```

Comment:
If any code in this block throws, we'll be left with an orphan lock in the DB which nobody knows about... not sure if we need to mitigate, but might be worth a thought..